### ISSUE TITLE
Dosbox-X mounts primary partition only if it has more than 1 partition

### ISSUE THREAD

**Issue by Siddhant-Mashalkar** [joncampbell123/dosbox-x#2678]
i was installing Windows 95 , i made a disk using bximage , then i created 1 PRI DOS partition and 1 EXT DOS Partition
when i imgmount the img , it only mounts PRI DOS parition
any way to mount EXT DOS partition?

**Comment by Wengier** [comment]
THE IMGMOUNT command only mounts one specified drive letter or number, so I do not think it supports multiple partitions in a single image at this time, although it may be possible to add support for such cases via some option(s).

**Comment by rderooy** [comment]
Indeed, this case is not supported by the integrated DOS. If you need multiple drives, it is recommended to create one disk image per drive letter and mount them separately. 

This is also documented on the wiki.

e.g.,
```
imgmount c c-drive.img
imgmount d d-drive.img
```

**Comment by joncampbell123** [comment]
You can however specify which partition (by index) using the -partidx mount option. i.e. -partidx 0

https://github.com/joncampbell123/dosbox-x/blob/master/src/dos/drive_fat.cpp#L1399

**Comment by Wengier** [comment]
Indeed, you can specify a partition to mount with "partidx" parameter, e.g.

``
imgmount c dos.img -o partidx=0
``

When I created an image with 2 partitions and tested it with "partidx=1" etc though, it failed with the error "Cannot create drive from file.". The logging for "partidx=1" says "Boot sector appears to have been created by FDISK.EXE but not formatted", and the logging for "partidx=2" says "Rejecting image, boot sector has weird values not consistent with FAT filesystem". Both partitions can be seen by booted DOS (6 or 7) though.

**Comment by Siddhant-Mashalkar** [comment]
formatted using DOS 6.22 but it actually couldn't create drive , why
![image](https://user-images.githubusercontent.com/71384075/125936154-06098d47-9937-4450-8b73-56d922c86d3c.png)

**Comment by Wengier** [comment]
@Siddhant-Mashalkar Please check out the log file and find out the exact error message(s). Perhaps similar to the messages I posted above.

**Comment by Siddhant-Mashalkar** [comment]
ok, the error message is
"Boot sector appears to have been created by FDISK.EXE but not formatted"
then i reformatted the Extended DOS Partition but it didn't work

**Comment by rderooy** [comment]
Yeah the support for extended partitions is probably simply incomplete/broken. Last time I looked at the partition code (to add support for auto detecting older DOS 2.0-3.21 partition types), I don't recall even seeing the code for extended partitions. In fact I was not even aware of the ``partidx`` option.

**Comment by joncampbell123** [comment]
Extended partitions are easy enough to support. Within the MBR partition marked extended, there is a sort of "linked list" of partition tables inside. The partition handing could be modified to use that.

To avoid copy-pasta for normal + extended partitions the code could be written to read the partition table entries into memory and then work from the in-memory list.

See also "DOS extended partitions" http://hackipedia.org/browse.cgi/Disk%20formats/Partition%20tables/MS%2dDOS/Partitions%20by%20Alex%20Verstak%20%281998%2d03%2d10%29%2ehtml

**Comment by joncampbell123** [comment]
See also: https://wiki.osdev.org/Partition_Table

**Comment by joncampbell123** [comment]
The partidx option is "-o partidx=*n*" where *n* is a number starting from 0 to select the partition to use.

The latest commit adds support for extended partitions.

**Comment by Wengier** [comment]
@joncampbell123 Yes, I can confirm with the latest code the parameter ``-o partidx=0`` works now. Great work!

**Comment by Wengier** [comment]
@rderooy You probably want to add information for mounting extended partitions in Wiki too. Thanks.

**Comment by rderooy** [comment]
@Wengier I updated the draft wiki. I did not test it myself, so I hope its fine. Here is what I added:

## Specifying the partition to mount
If the harddisk image contains multiple partitions, by default DOSBox-X will only mount the first primary partition.

You can optionally specify the ``-o partidx=0`` parameter, where 0 needs to be replaced by the partition you want to mount.
Zero is the same as not specifying the parameter as it represents the first primary partition.

e.g.

....
imgmount C: hdd.img -o partidx=1
....

NOTE: DOSBox-X only allows one partition per harddisk image to be mounted in the integrated DOS. When booting real DOS, all partitions will be available.

**Comment by joncampbell123** [comment]
@rderooy That is wrong, look at drive_fat.cpp. The partition index is zero based. Index 0 through 3 are mapped to the main partition table, whether there is something there or not. Indexes 4 and higher are mapped to extended partitions. The latest commit prints the partitions tables found as debug output for your use to the log file and/or console output.

drive_fat.cpp internally starts with partidx = -1 (meaning default choice) and assigns partidx to your value if provided. The code follows your partition index if partidx >= 0, and automatically picks otherwise.

**Comment by rderooy** [comment]
## Specifying the partition to mount
If the harddisk image contains multiple partitions, by default DOSBox-X will only mount the first primary partition.

You can optionally specify the ``-o partidx=`` parameter, where you can specify the partition number you want to mount.

- -1 means autodetect (default)
- 0 is the 1st primary partition
- 1 is the 2nd primary partition
- 2 is the 3rd primary partition
- 3 is the 4th primary partition or the Extended partition
- 4 is the 1st logical partition
- 5 is the 2nd logical partition
- 6 is the 3rd logical partition
- 7 is the 4th logical partition

e.g. to mount the 2nd primary partition:

....
imgmount C: hdd.img -o partidx=1
....

NOTE: DOSBox-X only allows one partition per harddisk image to be mounted in the integrated DOS. When booting real DOS, all partitions will be available.

NOTE: If the harddisk image has an Extended partition, that partition cannot be mounted, but the logical partitions it contains can.
